app/main.py

```diff
@@ -1,2 +1,136 @@
+from dataclasses import dataclass
+from typing import Optional, Any, List
+
+
+@dataclass
+class Node:
+    key: Any
+    value: Any
+    hash_value: int
+    next: Optional["Node"] = None # noqa
+
+
 class Dictionary:
-    pass
+    def __init__(
+            self,
+            initial_capacity: int = 8,
+            load_factor: float = 0.75
+    ) -> None:
+        self.capacity = initial_capacity
+        self.size = 0
+        self.load_factor = load_factor
+        self.table: List[Optional[Node]] = [None] * self.capacity
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        node_hash = hash(key)
+        index = self._hash(node_hash)
+        current_node = self.table[index]
+
+        if current_node is None:
+            self.table[index] = Node(key, value, node_hash)
+            self.size += 1
+        else:
+            while current_node:
+                if current_node.key == key:
+                    current_node.value = value
+                    return
+                if current_node.next is None:
+                    current_node.next = Node(key, value, node_hash)
+                    self.size += 1
+                    break
+                current_node = current_node.next
+
+        if self.size / self.capacity > self.load_factor:
+            self._resize()
+
+    def __getitem__(self, key: Any) -> Any:
+        node_hash = hash(key)
+        index = self._hash(node_hash)
+        current_node = self.table[index]
+
+        while current_node:
+            if current_node.key == key:
+                return current_node.value
+            current_node = current_node.next
+
+        raise KeyError(f"Key '{key}' not found")
+
+    def __len__(self) -> int:
+        return self.size
+
+    def _hash(self, node_hash: int) -> int:
+        return node_hash % self.capacity
+
+    def _resize(self) -> None:
+        old_table = self.table
+        self.capacity *= 2
+        self.table = [None] * self.capacity
+        self.size = 0
+
+        for current_node in old_table:
+            while current_node:
+                self[current_node.key] = current_node.value
+                current_node = current_node.next
```

Comment:
The `_resize` method is correctly implemented with O(n) time complexity in the best case and O(n^2) in the worst case due to collision handling, as specified in the checklist.